lib/writeUris.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+var fsExtra = require('fs-extra');
+var mime = require('mime');
+var path = require('path');
+var Promise = require('bluebird');
+
+var fsExtraOutputFile = Promise.promisify(fsExtra.outputFile);
+
+module.exports = writeUris;
+
+/**
+ * Write glTF resources as embedded data uris or external files.
+ *
+ * @param {Object} gltf The glTF asset.
+ * @param {String} gltfPath Path where the glTF will be saved.
+ * @param {Boolean} separateBuffers Writes out separate buffers.
+ * @param {Boolean} separateTextures Writes out separate textures.
+ * @param {Logger} logger A callback function for handling logged messages. Defaults to console.log.
+ * @returns {Promise} A promise that resolves to the glTF asset.
+ *
+ * @private
+ */
+function writeUris(gltf, gltfPath, separateBuffers, separateTextures, logger) {
+    var promises = [];
+
+    var buffer = gltf.buffers[Object.keys(gltf.buffers)[0]];
+    var bufferByteLength = buffer.extras._obj2gltf.source.length;
+
+    var texturesByteLength = 0;
+    var images = gltf.images;
+    for (var id in images) {
+        if (images.hasOwnProperty(id)) {
+            texturesByteLength += images[id].extras._obj2gltf.source.length;
+        }
+    }
+
+    // Buffers larger than ~192MB cannot be base64 encoded due to a NodeJS limitation. Source: https://github.com/nodejs/node/issues/4266
+    var exceedsMaximum = (texturesByteLength + bufferByteLength > 201326580);
+
+    if (exceedsMaximum) {
+        logger('Buffers and textures are too large to encode in the glTF, saving as separate resources.');
+    }
+
+    if (separateBuffers || exceedsMaximum) {
+        promises.push(writeSeparateBuffer(gltf, gltfPath));
+    } else {
+        writeEmbeddedBuffer(gltf);
+    }
+
+    if (separateTextures || exceedsMaximum) {
+        promises.push(writeSeparateTextures(gltf, gltfPath));
+    } else {
+        writeEmbeddedTextures(gltf);
+    }
+
+    return Promise.all(promises)
+        .then(function() {
+            deleteExtras(gltf);
+            return gltf;
+        });
+}
+
+function deleteExtras(gltf) {
+    var buffer = gltf.buffers[Object.keys(gltf.buffers)[0]];
+    delete buffer.extras;
+
+    var images = gltf.images;
+    for (var id in images) {
+        if (images.hasOwnProperty(id)) {
+            var image = images[id];
+            delete image.extras;
+        }
+    }
+}
+
+function writeSeparateBuffer(gltf, gltfPath) {
+    var buffer = gltf.buffers[Object.keys(gltf.buffers)[0]];
+    var source = buffer.extras._obj2gltf.source;
+    var bufferName = path.basename(gltfPath, path.extname(gltfPath));
+    var bufferUri = bufferName + '.bin';
+    buffer.uri = bufferUri;
+    var bufferPath = path.join(path.dirname(gltfPath), bufferUri);
+    return writeUris._outputFile(bufferPath, source);
+}
+
+function writeSeparateTextures(gltf, gltfPath) {
+    var promises = [];
+    var images = gltf.images;
+    for (var id in images) {
```

Comment:
You might be able to use [Promise.props](http://bluebirdjs.com/docs/api/promise.props.html) here instead of a manual for loop, but in both cases you have to be careful of the promise usage below because you could conceivably open too many files at once.